### DISCUSSION TITLE
Can you consider built-in deadpool to increase the stability of sqlx connection pool

### DISCUSSION THREAD

**Discussion by yuhaya** [SeaQL/sea-orm#1540] [General]
At present, the latest version of sqlx still has the problem of unstable connection pool. Under certain circumstances, the connection cannot be recovered and released normally, and abnormalities occur from time to time, which has a great impact on system stability.

[deadpool](https://github.com/bikeshedder/deadpool)

**Comment by tyt2y3** [comment]
To be fair, I think SQLx has a very good connection pool implementation, and the way it couples with transactions cannot be replicated with an external pool implementation.

If you have a reproducible case study, we can see what's the underlying problem.

**Reply by yuhaya** [reply]
This problem occurs randomly. In the past few days, I have tried unsuccessfully to reproduce it. If there is an example of successful reproduction in the future, I will post it.

**Reply by tyt2y3** [reply]
It may have seemed random, but it's usually something about async programming behind it.

**Reply by yuhaya** [reply]
The lack of the maximum number of file descriptors in linux is indeed not a problem with sqlx, but it would be better if the error message could be more clear @tyt2y3

**Comment by Sytten** [comment]
I am seeing quite frequent timeouts with the SQLx pool on postgres and I remember talking with the sqlx devs about the pool.
I think deadpool is more reliable in general. What do the transactions have to do with pooling exactly? I could see the argument for prepared statements, but not transactions.

**Reply by tyt2y3** [reply]
Do you have a link on the issue on sqlx pool's problem? Is it a design problem? Because when pools timeout, it means that they are saturated and so it can either 1) bail or 2) wait longer

**Reply by Sytten** [reply]
Its been a while, but I remember seeing issues where the requests where not served in order. So you could have a future that times out because it is never served even thought there is no contention issue. Also it seemed like connections sometimes are slow to return to the pool / are not returned sometimes.

IDK why it timeout, my pool is not exhausted and the DB is fine. So this smells like a pool issue to me. It is super hard to debug too since there is not much that bubbles up from sqlx. My timeout is at 5s and the db ping is like <10 ms.

I reduced the lifetime of connections, will see if it improves.
